### PR TITLE
Add "DeepStats" feature

### PR DESCRIPTION
**Pull request by jmichelsen**
DeepStats has several pieces to it right off the bat and more to come as I finish building out the API.

First, it gives us a way to have anyone running kolbot to automatically report SOJ sale so we can all share in that data. This would make DClone hunting _possible_ again. SOJ reports and previous walks can be found at https://gheedsfortune.com/diablo-clone-hunt/

Second, the Item Reporting feature reports anything your bot _keeps_. The items will show up in the "Global Inventory" at https://gheedsfortune.com and also if you log in, you will be able to see just your items in a similar feed (from anywhere! Check your drops on the go :) )

On both of these pages are stats collected and displayed about bots that have reported. Things like Average MF, location, char class, etc.

The Search functionality on both the Global Inventory and Your Inventory narrows down both the items displayed, and the stats, giving you the ability to look at details surrounding a single item. For example, you can search "stone of jordan" and see that the majority of SOJs came from Catacombs Level 4 (Andy).

Along with item reporting via API, DeepStats data is dumped to a new directory `kolbot/data/deepstats/{yourProfile}.dump.json` so if you don't want to use any of the APIs but still want the data, it's available to you by using the "DeepStats.FileLogOnly"

Finally, Alerts can be created around both SOJ sales for Dclone hunting, or around item drops reported by your bot. So if you're waiting on a specific item, you can be alerted as soon as that item is found via sms, twitter DM, or discord PM. The alerting is still a work in progress but should be done soon.

Big thanks to Jaenster for whipping together the HTTP module and poirier for helping troubleshoot the d2bs websocket stuff!

### CODE REVIEW

**Comment by Fa-b** [comment]
Sounds really nice! One question though, how is a Gheedsfortune Login connected to the bots I‘m running? I would expect the logging to be anonymous?

**Comment by jaenster** [comment]
I like the idea of counting sojs.

Im not so much for counting other things, or link it to an account. Just my 2 cents

Im undecided on this one for now

**Comment by jmichelsen** [comment]
The auth token and account lets you link _your_ items to your account, and enables the other features like alerts (and more to come). If the data is anonymous, _you_ can't act on it.

**Comment by jaenster** [comment]
> 
> 
> The auth token and account lets you link _your_ items to your account, and enables the other features like alerts (and more to come). If the data is anonymous, _you_ can't act on it.

If we are talking about an anonymous soj counter, it kinda doesn't matter.

But, yeah like i said i dont know what i think about this yet

**Comment by jmichelsen** [comment]
SOJ counting could certainly be anonymous (despite security concerns about having an unauthenticated POST api on the net)

Also, the API is public so you can check out the raw data and do things with it yourself: https://gheedsfortune.com/api/v1/

I'm building an Android app on top of the API, for example. More options to keep tabs on your bots when you're away from your computer.

**Comment by Fa-b** [comment]
I like the idea of counting everything. Just for statistics. But i prefer it anonymous! I don‘t need a „one tool for all tasks“ solution. Its a matter of security

**Comment by jmichelsen** [comment]
Fair enough. I can understand that @Fa-b. I could modify it so the token is optional, and have a separate endpoint for anonymous stats. What do ya think of that?

**Comment by Fa-b** [comment]
Would that change the data pushed from kolbot to gheedsfortune? I don‘t think so..

**Comment by jmichelsen** [comment]
I'm not sure I understand what the concern is. The data that's being sent is in this change set, you can see what's being sent. Account name is not being sent. Character name is, but only so I can build stats on how many individual characters are reporting data. That could be a hash instead of a real name though, if the concern is that "character name" isn't anonymous.

**Comment by jmichelsen** [comment]
All the URLs in this changeset point to gheedsfortune but they are configurable if anyone wanted to build their own backend. The data is just json. I could yank everything specific to gheedsfortune out and just leave the config settings empty as a "template" for posting data to any backend, and a wiki about how to configure it to gheedsfortune specifically or something.

I built gheeds because I build backends. That's what I'm good at haha. I felt the need in my own botting for a central place to collect stats, share that data with others, and do stuff that can't be done inside kolbot like integrate with other technologies and services on the Internet. It's all opt-in...

**Comment by Fa-b** [comment]
> That could be a hash instead of a real name though, if the concern is that "character name" isn't anonymous
I think It would be fine then

**Comment by mf022** [comment]
Did you think to add a DeepStats.md for these, maybe in the [d2tools](https://github.com/blizzhackers/d2tools) repository?

Are you also interested to develop pro features on your website?

File: d2bs/kolbot/libs/common/Misc.js [hunk 1]
```diff
@@ -1406,6 +1407,7 @@ var Misc = {
 
 		if (action.match("kept", "i")) {
 			lastArea = DataFile.getStats().lastArea;
+			lastAreaID = DataFile.getStats().lastAreaID;
 
 			if (lastArea) {
```
Comment (mf022) [inline d2bs/kolbot/libs/common/Misc.js]: lastArea = ?

**Comment by laztheripper** [comment]
No feature from the master branch should depend on someone's external site. That's my 2cents, noah told me as much when I offered something similar.

**Comment by jmichelsen** [comment]
I'm going to close this Pr and open one with just the skeleton of this, in case anyone wants to collect the data to files themselves, or make arbitrary http requests. I'll keep the rest of the functionality in my fork. Thanks for the feedback.

**Comment by jmichelsen** [comment]
Jan wrote the HTTP and Sockets modules, so I'll submit after he gets those in

**Comment by ehgit** [comment]
I don't know about WhatsApp, but think this is easier done with a telegram bot - if there are several other people up for that feature, I might give it a shot.

**Pull request closed without merging**